packages/forms/src/index.tsx

```diff
@@ -211,6 +312,12 @@ const setCoercion = (
     // This is for handling optional relation id fields, like a text input for
     // `userId` if the user relation is optional
     validation.setValueAs = (val: string) => val || undefined
```

Comment:
We need to make it possible to override this behavior with `emptyAsUndefined={false}`

Reply:
@Tobbe, I want to make sure I'm understanding you clearly, identify a concern, and propose some alternatives.  The proposed interface is:

**Case 1:**
If the user does not set `emptyAsUndefined`,  the `emptyAsUndefined ` prop would be `undefined` and the name ending in the `Id` test is applied.  The field returns `undefined` if the name ends in `Id` and is empty.

**Case 2:**
If the user does set `emptyAsUndefined`,  the `emptyAsUndefined`  prop will be `true `and empty values will be returned as `undefined` regardless of what the name of the field is.

**Case 3:**
If the user does set `emptyAsUndefined={false}`,  the `emptyAsUndefined` prop will be `false`.  The field will return empty values as an empty string regardless of an `Id` in the name of the field.

Have I understood the proposed interface correctly?

The concern is that I believe this is confusing to users as we have a three-state boolean prop.  **I believe users typically expect boolean props to have two discrete states ... true or false.  I personally dislike three-state boolean props.**

**Potential Solutions:**
1.  Get rid of the `Id `check case and leave the `emptyAsUndefined` prop to handle this case.

2.  Create a new prop called `emptyAsUndefinedForNamesWithId`.  The prop could default to `true`, and optionally be set to false.  While simple to implement, the challenge is that we need to define precedence between the following three props, as there are numerous combinations of how they could potentially be applied:
  `emptyAsNull`
  `emptyAsUndefined`
  `emptyAsUndefinedForNamesWithId`

Given the multitude of cases, I personally don't love this alternative.

3.  Create an `enum` prop that defines how to handle empty fields.  This gets rid of the combination and precedence issues in alternative 2.  The enum would potentially have the following values:
  ` EmptyAs.NORMAL`
   `EmptyAs.UNDEFINED`
  ` EmptyAs.NULL`
   `EmptyAs.UNDEFINED_FOR_NAMES_WITH_ID`

Thoughts?

Reply:
> We need to make it possible to override this behavior with `emptyAsUndefined={false}`

meaning this?

```js
value ?? undefined
```

Reply:
I agree that tri-state booleans aren't great. Unfortunately I don't think there's a way around that seeing how html attributes work in React. There is always going to be three states: 1) `boolAttribute={true}` (or simply just `boolAttribute`) 2) `boolAttribute={false}` and 3) Attribute not specified at all. What we can do to ease the cognitive load is make two of those states behave the same. And I think `emptyAsUndefined` satisfies that.

### Text fields and select fields with names ending in `Id`
**State 1:**
`emptyAsUndefined={true}`. Same as leaving it out. Default behavior. The field will return `undefined` for the value`''`

**State 2:**
`emptyAsUndefined={false}`. The field will return `''` for the value `''`

**State 3:**
Same as State 1

### Other fields
**State 1:**
`emptyAsUndefined={true}`. The field will return `undefined` for the value `''`

**State 2:**
`emptyAsUndefined={false}`. Same as leaving it out. Default behavior. The field will return `''` for the value `''`

**State 3:**
Same as State 2

That's how I saw it when I thought about the implementation.

Reply:
OK, the other way we can look at it is that the default value of `emptyAsUndefined `will depend upon whether the name ends in `Id`.

If this is what is desired, I'll gladly code it up (Even if I believe the interface is confusing 😊).

Thanks for clarifying Tobbe!

Reply:
> > We need to make it possible to override this behavior with `emptyAsUndefined={false}`
> 
> meaning this?
> 
> ```js
> value ?? undefined
> ```

Hi @jonschlinkert, thanks for dropping in 🙂 

Yes, with `emptyAsUndefined={false}` we could switch to `??` as that doesn't treat `''` as a false value. The bigger issue at hand though, with the code as it currently sits, is that we don't look at `emptyAsUndefined` at all here.

Reply:
@Tobbe, sorry for the delay!  (and thanks for the feedback!)

OK, I'll make one more attempt at a single prop solution!  I'm proposing the following:

**1.  A prop called `emptyAs`**

The type of the `emptyAs` props would be `type EmptyAsValue = null | 'undefined' | 0 | ''`

Example usage:
```
        <TextField name="textField" emptyAs={'undefined'} />
        <SelectField
          name="selectField"
          defaultValue=""
          emptyAs={null}
        >
```

It's simple as we tell it exactly what type we want to return on empty.  It's also easy to extend if we ever need to.  The `undefined` case being a string `"undefined"` is a necessary deviation, and passing an actual defined would mean that we would always have to default to `undefined`, which we wouldn't want to do.

Thoughts?

**2.  But what about the case of a required input that is a relation (ends in `Id`)?**

I'm sorry, but the more I think about it, the more I think that the current code does not really do what we always want.  And I will take some responsibility as I originally pushed for and coded an `undefined` case if not required!

Let's go through the example right from the test, assuming `groupID` is a relation field:

```
        <SelectField name="groupId" defaultValue="">
          <option value="">No group</option>
          <option value={1}>Group 1</option>
          <option value={2}>Group 2</option>
        </SelectField>
```

The current code will send back `undefined` if "No group" is selected, causing the API service and Prisma to not try to write a value to the relation field.  If the relation field is optional (likely if the field is not required), then a null remains in the database groupID field and the system remains in a good state.

However, let's say that the system has `groupID` set to `2` (Group 2).  It's still an optional field.  The user goes to the form, and the input's default value gets set to `2` (Group 2).  Now the user wants to deselect `2` (Group 2) and assign (No Group).  After selecting (No Group), the field will return `undefined` for `groupId`.  The `undefined` field will be ignored by Prisma (unless there is special code to handle), and the value for the `groupId` will not have been changed as desired.

My recommendation is that we allow the programmer to decide what the appropriate return value for empty cases is.  For a lot of relations, it may be `null`.

Thoughts?

Reply:
> The undefined case being a string "undefined" is a necessary deviation, and passing an actual defined would mean that we would always have to default to undefined, which we wouldn't want to do.

So, what would be the default? (I think I know the answer, but please answer so we know we're both on the same page here)

> The undefined field will be ignored by Prisma (unless there is special code to handle), and the value for the groupId will not have been changed as desired.

Being able to go back to the default "empty" state is super important. Are you saying `null` would be a better default value for not-required, ends-in-Id fields?

> My recommendation is that we allow the programmer to decide what the appropriate return value for empty cases is.

Absolutely agree that we should allow the programmer to decide. And I think `emptyAs` does that 🙂  But I'm not convinced we should force the programmer to decide (if that's what you were implying. Sorry if not)

Reply:
> Once again, my apologies for being so late to respond! I'm hoping my availability should improve shortly.

It's all good. I'm just happy for all the time you've been able to spend on this

> However, `NumberFields` and `SelectFields` could be used for relations

And so could `TextField`s. Remember that the Issue that started all of this was [Scaffolds don't handle optional relations with string keys #4297](https://github.com/redwoodjs/redwood/issues/4297). And because of that I'd really like us to continue supporting the ends-with-Id case.

> Yes, I believe so. Do you see any concerns with this?

I don't have enough experience to know. I'll defer to you on this.